src/Microsoft.PowerShell.Commands.Utility/commands/utility/CsvCommands.cs

```diff
@@ -1336,35 +1336,32 @@ internal void ReadHeader()
 
             var values = new List<string>(ValueCountGuestimate);
             var builder = new StringBuilder(LineLengthGuestimate);
+
             while ((Header == null) && (!this.EOF))
             {
-                ParseNextRecord(values, builder);
-
-                // Trim all trailing blankspaces and delimiters ( single/multiple ).
-                // If there is only one element in the row and if its a blankspace we dont trim it.
-                // A trailing delimiter is represented as a blankspace while being added to result collection
-                // which is getting trimmed along with blankspaces supplied through the CSV in the below loop.
-                while (values.Count > 1 && values[values.Count - 1].Equals(string.Empty))
+                if (PeekNextChar('#'))
                 {
-                    values.RemoveAt(values.Count - 1);
-                }
+                    ParseNextRecord(values, builder);
+                    TrimEmptiesAtEnd(values);
 
-                // File starts with '#' and contains '#Fields:' is W3C Extended Log File Format
-                if (values.Count != 0 && values[0].StartsWith("#Fields: "))
-                {
-                    values[0] = values[0].Substring(9);
-                    Header = values;
-                }
-                else if (values.Count != 0 && values[0].StartsWith('#'))
-                {
-                    // Skip all lines starting with '#'
-                }
-                else
-                {
-                    // This is not W3C Extended Log File Format
-                    // By default first line is Header
-                    Header = values;
+                    if (values.Count != 0 && values[0].StartsWith("#Fields: "))
```

Comment:
We parse record and then do a check for `#Fields:` but we can do the check and only then parse if needed.